src/app/clusters/webrtc-transport-provider-server/webrtc-transport-provider-server.cpp

```diff
@@ -0,0 +1,501 @@
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Protocols::InteractionModel::Status::Failure);
+        return;
+    }
+
+    switch (ctx.mRequestPath.mCommandId)
+    {
+    case Commands::SolicitOffer::Id:
+        CommandHandlerInterface::HandleCommand<Commands::SolicitOffer::DecodableType>(
+            ctx, [this](HandlerContext & subCtx, const auto & req) { HandleSolicitOffer(subCtx, req); });
+        break;
+
+    case Commands::ProvideOffer::Id:
+        CommandHandlerInterface::HandleCommand<Commands::ProvideOffer::DecodableType>(
+            ctx, [this](HandlerContext & subCtx, const auto & req) { HandleProvideOffer(subCtx, req); });
+        break;
+
+    case Commands::ProvideAnswer::Id:
+        CommandHandlerInterface::HandleCommand<Commands::ProvideAnswer::DecodableType>(
+            ctx, [this](HandlerContext & subCtx, const auto & req) { HandleProvideAnswer(subCtx, req); });
+        break;
+
+    case Commands::ProvideICECandidates::Id:
+        CommandHandlerInterface::HandleCommand<Commands::ProvideICECandidates::DecodableType>(
+            ctx, [this](HandlerContext & subCtx, const auto & req) { HandleProvideICECandidates(subCtx, req); });
+        break;
+
+    case Commands::EndSession::Id:
+        CommandHandlerInterface::HandleCommand<Commands::EndSession::DecodableType>(
+            ctx, [this](HandlerContext & subCtx, const auto & req) { HandleEndSession(subCtx, req); });
+        break;
+
+    default:
+        // Mark unrecognized command as UnsupportedCommand
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::UnsupportedCommand);
+        break;
+    }
+}
+
+// Helper functions
+WebRTCSessionTypeStruct * WebRTCTransportProviderServer::FindSession(uint16_t sessionId)
+{
+    for (auto & session : mCurrentSessions)
+    {
+        if (session.id == sessionId)
+        {
+            return &session;
+        }
+    }
+
+    return nullptr;
+}
+
+WebRTCTransportProviderServer::UpsertResultEnum
+WebRTCTransportProviderServer::UpsertSession(const WebRTCSessionTypeStruct & session)
+{
+    // Use std::find_if to locate a matching session ID in mCurrentSessions
+    auto it = std::find_if(mCurrentSessions.begin(), mCurrentSessions.end(),
+                           [id = session.id](const auto & existing) { return existing.id == id; });
+
+    if (it != mCurrentSessions.end())
+    {
+        // Found an existing session with the same ID; update it
+        *it = session;
+
+        MatterReportingAttributeChangeCallback(AttributeAccessInterface::GetEndpointId().Value(), WebRTCTransportProvider::Id,
+                                               WebRTCTransportProvider::Attributes::CurrentSessions::Id);
+
+        return UpsertResultEnum::kUpdated;
+    }
+
+    // If not found, insert a new entry
+    mCurrentSessions.push_back(session);
+
+    MatterReportingAttributeChangeCallback(AttributeAccessInterface::GetEndpointId().Value(), WebRTCTransportProvider::Id,
```

Comment:
Can't thus just happen up front in only one place, if it happens on all codepaths through this function?